File: libs/compiler/src/main/java/cn/lliiooll/processors/FunctionItemProcessor.java
/*
 * QNotified - An Xposed module for QQ/TIM
 * Copyright (C) 2019-2021 dmca@ioctl.cc
 * https://github.com/ferredoxin/QNotified
 *
 * This software is non-free but opensource software: you can redistribute it
 * and/or modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or any later version and our eula as published
 * by ferredoxin.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * and eula along with this software.  If not, see
 * <https://www.gnu.org/licenses/>
 * <https://github.com/ferredoxin/QNotified/blob/master/LICENSE.md>.
 */

package cn.lliiooll.processors;

import com.google.auto.service.AutoService;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Set;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import nil.nadph.qnotified.base.annotation.FunctionEntry;

@AutoService(Processor.class)
@SupportedAnnotationTypes("nil.nadph.qnotified.base.annotation.FunctionEntry")
@SupportedSourceVersion(SourceVersion.RELEASE_8)
public class FunctionItemProcessor extends BaseProcessor {

    private TypeMirror iProvider = null;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        System.out.println(">>>> FunctionItemProcessor init <<<<");
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        Set<? extends Element> annos = roundEnv.getElementsAnnotatedWith(FunctionEntry.class);
        if (!annos.isEmpty()) {
            System.out.println(">>>> FunctionEntry Processing <<<<");
            ClassName absHook = ClassName.get("nil.nadph.qnotified.hook", "AbsDelayableHook");
            ClassName list = ClassName.get("java.util", "List");
            MethodSpec.Builder beyond = MethodSpec
                .methodBuilder("getAnnotatedFunctionItemClassList")
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                .returns(ParameterizedTypeName.get(list, absHook))
                .addStatement("$T result = new $T<>()", ParameterizedTypeName.get(list, absHook),
                    ArrayList.class);

            for (Element e : annos) {
//                System.out.println("Processing >>> " + e.toString());
                beyond.addStatement("result.add($T.INSTANCE)", e.asType());
            }
            beyond.addStatement("return result");
            TypeSpec util = TypeSpec.classBuilder("AnnotatedFunctionItemList")
                .addModifiers(Modifier.PUBLIC)
                .addMethod(beyond.build())
                .build();
            JavaFile javaFile = JavaFile.builder("nil.nadph.qnotified.gen", util)
                .build();
            try {
                javaFile.writeTo(env.getFiler());
            } catch (IOException e) {
                System.out.println(e.getMessage());
            }
            System.out.println(">>>> FunctionEntry Processed(total: " + annos.size() + ") <<<<");
        } else {
            System.out.println(">>>> FunctionEntry skipped because empty <<<<");
        }
        return false;
    }
}

File: libs/compiler/src/main/kotlin/cn/lliiooll/processors/FunctionItemProcessor.kt
/*
 * QNotified - An Xposed module for QQ/TIM
 * Copyright (C) 2019-2022 dmca@ioctl.cc
 * https://github.com/ferredoxin/QNotified
 *
 * This software is non-free but opensource software: you can redistribute it
 * and/or modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or any later version and our eula as published
 * by ferredoxin.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * and eula along with this software.  If not, see
 * <https://www.gnu.org/licenses/>
 * <https://github.com/ferredoxin/QNotified/blob/master/LICENSE.md>.
 */

package cn.lliiooll.processors

import com.google.devtools.ksp.processing.*
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ksp.KotlinPoetKspPreview
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.writeTo

@KotlinPoetKspPreview
class FunctionItemProcessor(private val codeGenerator: CodeGenerator, private val logger: KSPLogger) :
    SymbolProcessor {

    override fun process(resolver: Resolver): List<KSAnnotated> {
        val symbols =
            resolver.getSymbolsWithAnnotation("nil.nadph.qnotified.base.annotation.FunctionEntry")
                .filter { it is KSClassDeclaration }
                .map { it as KSClassDeclaration }
                .toList()
        if (symbols.isEmpty()) {
            logger.warn(">>>> FunctionEntry skipped because empty <<<<")
            return emptyList()
        }

        logger.warn(">>>> FunctionEntry Processing <<<<")
        val simpleNameMap = HashMap<String, String>(symbols.size)
        val mGetApi = FunSpec.builder("getAnnotatedFunctionItemClassList").run {
            addCode(CodeBlock.Builder().run {
                add("return arrayOf(«")
                symbols.forEachIndexed { index, ksClassDeclaration ->
                    if (simpleNameMap.contains(ksClassDeclaration.simpleName.asString())) {
                        logger.error("Duplicate name in FunctionEntry's simpleName: ${ksClassDeclaration.qualifiedName?.asString()?:"null"}, ${simpleNameMap[ksClassDeclaration.simpleName.asString()]}")
                    } else {
                        simpleNameMap[ksClassDeclaration.simpleName.asString()] = ksClassDeclaration.qualifiedName?.asString()
                            ?: "null"
                    }
                    val isJava = ksClassDeclaration.containingFile?.filePath?.endsWith(".java") == true
                    // logger.warn("Processing >>> $ksClassDeclaration,isJava = $isJava")
                    val typeName = ksClassDeclaration.asStarProjectedType().toTypeName()
                    val format = StringBuilder("\n%T").run {
                        if (isJava) append(".INSTANCE")
                        if (index == symbols.lastIndex) append("\n") else append(",")
                        toString()
                    }
                    add(format, typeName)
                }
                add("»)")
                build()
            })
            build()
        }
        logger.warn(">>>> FunctionEntry Processed(total: " + symbols.size + ") <<<<")
        // @file:JvmName("AnnotatedFunctionItemList")
        val annotationSpec = AnnotationSpec.builder(JvmName::class).run {
            addMember("%S", "AnnotatedFunctionItemList")
            build()
        }
        val dependencies = Dependencies(true, *(symbols.map {
            it.containingFile!!
        }.toTypedArray()))
        FileSpec.builder("nil.nadph.qnotified.gen", "AnnotatedFunctionItemList")
            .addAnnotation(annotationSpec)
            .addFunction(mGetApi)
            .build()
            .writeTo(codeGenerator, dependencies)
        return emptyList()
    }
}

@KotlinPoetKspPreview
class FunctionItemProvider : SymbolProcessorProvider {
    override fun create(
        environment: SymbolProcessorEnvironment
    ): SymbolProcessor {
        return FunctionItemProcessor(environment.codeGenerator, environment.logger)
    }
}
